### PR TITLE
Add-dotnet-payment-service

### PR DESCRIPTION
**Pull request by ulili5**
*Issue #42, if available:*

*Description of changes:*

- Add the .NET Payment Services
- Eddited the API Gateway to Include the Proxy to the downstream Payment Services
- Updated the Front End (Angular) to include Form for Payment (add/edit)
- Updated the PetForm to include a Grid of payments

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by zhenweig — APPROVED** [review]
Does the [overall README file](https://github.com/aws-observability/application-signals-demo/blob/main/README.md) also need updates with the extension, such as prerequisites and deploy scripts?

**Comment by ulili5** [comment]
> Does the [overall README file](https://github.com/aws-observability/application-signals-demo/blob/main/README.md) also need updates with the extension, such as prerequisites and deploy scripts?

Probably no, because I was able to follow this instruction to deploy using shell, the team might need to review the other tow options (Terraform and EC2)

**Review by AsakerMohd — COMMENTED** [review]
I've added another commit after testing the EKS/EC2 scripts as well as updated the traffic generator. Was able to confirm that both EKS and EC2 deployments are working. For Terraform, it uses the same images and steps as the EKS steps and for K8s, it also uses the same images. Other than that, this should be ready to merge for adding a new payment dotnet service.

File: dotnet-petclinic-payment/PetClinic.PaymentService/appsettings.Development.json [hunk 1]
```diff
@@ -0,0 +1,9 @@
+{
+  "Logging": {
+    "LogLevel": {
+      "Default": "Information",
+      "Microsoft.AspNetCore": "Warning"
+    }
+  },
+  "PetClinicBaseAddress": "http://ec2-3-106-218-72.ap-southeast-2.compute.amazonaws.com:8080/"
```
Comment (pxaws) [inline dotnet-petclinic-payment/PetClinic.PaymentService/appsettings.Development.json:8]: Why we are using this hardcoded dns name `http://ec2-3-106-218-72.ap-southeast-2.compute.amazonaws.com`?

File: dotnet-petclinic-payment/PetClinic.PaymentService/readme.md [hunk 2]
```diff
@@ -0,0 +1,7 @@
+# Local Dev Guide
+
+The PetClinic uses Eureka Services discovery, to test this application locally you need to run Eureka Server, use the following command to run it using docker
+
+```
+ docker run -d --publish 8761:8761 public.ecr.aws/o9d5t0i3/eureka-server:latest
```
Comment (pxaws) [inline dotnet-petclinic-payment/PetClinic.PaymentService/readme.md:6]: I don't think it's reasonable to use a repo from non-verified account like this `public.ecr.aws/o9d5t0i3/eureka-server`. Can we give instruction for customers to build the image themselves?

File: scripts/eks/appsignals/deploy-sample-app.sh [hunk 3]
```diff
@@ -33,6 +33,13 @@ if [[ $OPERATION == "apply" ]]; then
         --attach-policy-arn arn:aws:iam::aws:policy/AmazonBedrockFullAccess \
         --approve \
         --override-existing-serviceaccounts
+
+    aws dynamodb create-table \
```
Comment (pxaws) [inline scripts/eks/appsignals/deploy-sample-app.sh:37]: Creating the dynamodb table here is strange here and error prone. What if the table is already exist? We usually let the sample app to create its own resources so that it's self contained. 

Can we do it inside the payment service for consistency? Like what the java services do, we want to create the table only if it doesn't exist.

File: scripts/eks/appsignals/tf-deploy-k8s-res.sh [hunk 4]
```diff
@@ -1,7 +1,7 @@
 #!/bin/bash
 
 CLUSTER_NAME=${1:-"python-apm-demo"}
-REGION=${2:-"us-east-1"}
+REGION=${2:-"us-west-2"}
```
Comment (pxaws) [inline scripts/eks/appsignals/tf-deploy-k8s-res.sh:4]: Why change the version here?

File: spring-petclinic-api-gateway/src/main/resources/static/scripts/pet-form/pet-form.template.html [hunk 5]
```diff
@@ -4,7 +4,7 @@ <h2>Pet</h2>
     <div class="form-group">
         <label class="col-sm-2 control-label">Owner</label>
         <div class="col-sm-6">
-            <p class="form-control-static">{{$ctrl.pet.owner}}</p>
+            <p class="form-control-static">{{$ctrl.pet.owner.name}}</p>
```
Comment (pxaws) [inline spring-petclinic-api-gateway/src/main/resources/static/scripts/pet-form/pet-form.template.html:7]: I guess owner will have other properties than `name`. What's the reason for the change here?

File: traffic-generator/index.js [hunk 6]
```diff
@@ -159,4 +168,43 @@ for (let i = 0; i < 2; i++) {
...
+        axios.get(`${baseUrl}/api/payments/owners/1/pets/1`, { timeout: 10000 })
+            .catch(err => {
+                console.error(`${baseUrl}/api/payments/owners/1/pets/1, error: ` + (err.response ? err.response.data : err.toString()));
+            }); // Catch and log errors
+    }
+}, { scheduled: false });
+
+lowTrafficPaymentTask.start();
+
+const generateHighPaymentLoad = async () => {
```
Comment (pxaws) [inline traffic-generator/index.js:192]: Suggest to remove the high payment load. We don't need this. It will generate a lot of traffic which incurs more cost.

File: dotnet-petclinic-payment/PetClinic.PaymentService/appsettings.Development.json (hunk 1 above)
Comment (AsakerMohd) [reply dotnet-petclinic-payment/PetClinic.PaymentService/appsettings.Development.json:8]: removed

File: dotnet-petclinic-payment/PetClinic.PaymentService/readme.md (hunk 2 above)
Comment (AsakerMohd) [reply dotnet-petclinic-payment/PetClinic.PaymentService/readme.md]: removed as it's not needed.

File: scripts/eks/appsignals/deploy-sample-app.sh (hunk 3 above)
Comment (AsakerMohd) [reply scripts/eks/appsignals/deploy-sample-app.sh]: I added logic to check if the table exists before creating it.

File: scripts/eks/appsignals/tf-deploy-k8s-res.sh (hunk 4 above)
Comment (AsakerMohd) [reply scripts/eks/appsignals/tf-deploy-k8s-res.sh]: reverted

File: traffic-generator/index.js (hunk 6 above)
Comment (AsakerMohd) [reply traffic-generator/index.js]: removed.

File: spring-petclinic-api-gateway/src/main/resources/static/scripts/pet-form/pet-form.template.html (hunk 5 above)
Comment (AsakerMohd) [reply spring-petclinic-api-gateway/src/main/resources/static/scripts/pet-form/pet-form.template.html]: reverted

**Comment by pxaws** [comment]
1. We need to create the ddb table inside the sample app rather than relying on bash script to create it
2. We need to have a way to clean up the ddb payment entries produced by the traffic generator periodically (maybe once every day). We don't want the payment entries to increase without limit, otherwise it will overwhelm the service eventually. 

Other than the above 2 points, the rest of the code looks good to me.

**Comment by AsakerMohd** [comment]
> 1. We need to create the ddb table inside the sample app rather than relying on bash script to create it
> 
>     2. We need to have a way to clean up the ddb payment entries produced by the traffic generator periodically (maybe once every day). We don't want the payment entries to increase without limit, otherwise it will overwhelm the service eventually.
> 
> 
> Other than the above 2 points, the rest of the code looks good to me.

1. Code updated so the DDB table is created during application startup if it doesn't exist. 
2. A new API is added that basically empties the DDB and that API is called by the traffic generator every 6 hours.

File: dotnet-petclinic-payment/PetClinic.PaymentService/PetClinicContext.cs [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+    {
+        try
+        {
+            if (Attribute.GetCustomAttribute(typeof(Payment), typeof(DynamoDBTableAttribute)) is DynamoDBTableAttribute att)
+            {
+                var request = new DeleteTableRequest
+                {
+                    TableName = att.TableName
+                };
+                await AmazonDynamoDBClient.DeleteTableAsync(request);
```
Comment (pxaws) [inline dotnet-petclinic-payment/PetClinic.PaymentService/PetClinicContext.cs:114]: Will it delete the table completely or just clear the content of the table?

**Review by pxaws — APPROVED** [review]
